src/codegen/riscv64/macro-assembler-riscv64.cc

```diff
@@ -1466,7 +1466,6 @@ void TurboAssembler::li(Register rd, Operand j, LiFlags mode) {
   } else if (mode == ADDRESS_LOAD) {
     // We always need the same number of instructions as we may need to patch
     // this code to load another value which may need all 8 instructions.
-    RecordRelocInfo(RelocInfo::INTERNAL_REFERENCE_ENCODED);
```

Comment:
The RecordRelocInfo is already set by the caller and rest of the code expects the value the caller set. The solution may be to set  RecordRelocInfo(RelocInfo::INTERNAL_REFERENCE_ENCODED) in the caller of li. WE cannot change the test because it is common to all targets. 

I prefer you do it, because I do not have access to the hardware and this test as you say fails in hardware. It is also interesting to know why fails in hardware and not in the simulator! The other question is why this test is put under DEBUG!

Reply:
> The RecordRelocInfo is already set by the caller and rest of the code expects the value the caller set. The solution may be to set RecordRelocInfo(RelocInfo::INTERNAL_REFERENCE_ENCODED) in the caller of li. WE cannot change the test because it is common to all targets.

I intend to fix it.

> It is also interesting to know why fails in hardware and not in the simulator


The problem is when read a instr from mem where a piece of assembler is copied into, x86 is normal but riscv64 report sigsegv. These assembler:
```
   0x2aaf2f87dc:        nop
   0x2aaf2f87e0:        nop
   0x2aaf2f87e4:        ret
   0x2aaf2f87e8:        nop
   0x2aaf2f87ec:        lui     a4,0x0
   0x2aaf2f87f0:        addiw   a4,a4,3
   0x2aaf2f87f4:        slli    a4,a4,0xc
   0x2aaf2f87f8:        addi    a4,a4,-1361
   0x2aaf2f87fc:        slli    a4,a4,0xc
   0x2aaf2f8800:        addi    a4,a4,760
   0x2aaf2f8804:        slli    a4,a4,0xc
   0x2aaf2f8808:        addi    a4,a4,2012
   0x2aaf2f880c:        jr      a4   ->   0x2aaf2f87dc
   0x2aaf2f8810:        nop
   0x2aaf2f8814:        ebreak
```
It is copyed to 0x3fee582044:
```
   0x3fee582044:        or      a4,a0,zero
   0x3fee582048:        j       0x3fee58205c
   0x3fee58204c:        nop
   0x3fee582050:        nop
   0x3fee582054:        ret
   0x3fee582058:        nop
   0x3fee58205c:        lui     a4,0x0
   0x3fee582060:        addiw   a4,a4,3
   0x3fee582064:        slli    a4,a4,0xc
   0x3fee582068:        addi    a4,a4,-1361
   0x3fee58206c:        slli    a4,a4,0xc
   0x3fee582070:        addi    a4,a4,760
   0x3fee582074:        slli    a4,a4,0xc
   0x3fee582078:        addi    a4,a4,2012
   0x3fee58207c:        jr      a4 ->   0x2aaf2f87dc
```
There are 3 different that may be caused it :

- different linux kernel version, hardware(5.5) vs simulator(4.19). But it also failed when test simulator on linux 5.3.
- different complier. GCC vs Clang
- different execution way. On  simulator, it is readed from mem and explain execution. But on riscv64 hardware, it is execution by cpu.

I think the third point is the most likely, but I still have no way to prove this.

>

Reply:
One thing I see in the above code is the branch to a fixed address. That address might be valid in simulator, but nonexistent in the hardware. We should make sure  that that address already exists and you have that much memory in your hardware. What are we doing here? What the goal is? What is the statement in high level language that generates this code? May be we are allocating too much. May be the heap space is colliding with the stack space. In your test make the allocated space smaller and see if it still fails.

Reply:
This code is not generate by js.It is a test case for test the LoadAddress, see: https://github.com/v8-riscv/v8/blob/riscv-porting-dev/test/cctest/test-macro-assembler-riscv64.cc#L115
```
  __ mov(a4, a0);

  __ Branch(&skip);
  __ bind(&to_jump);
  __ nop();
  __ nop();
  __ jr(ra);
  __ nop();
  __ bind(&skip);
  __ li(a4, Operand(masm->jump_address(&to_jump)), ADDRESS_LOAD);
  int check_size = masm->InstructionsGeneratedSince(&skip);
  // NOTE (RISCV): current li generates 8 instructions, if the sequence is
  // changed, need to adjust the CHECK_EQ value too
  CHECK_EQ(8, check_size);
  __ jr(a4);
  __ nop();
  __ stop();
  __ stop();
  __ stop();
  __ stop();
  __ stop();
```
These code generates these assemblers.
And my hardware has 8GB RAM.

Reply:
I find reason:
0x2aaf2f87dc is buffer of assembler which is NOT_EXECUTABLE, but 0x3fee582048 is code-space that is EXECUTABLE.
So wo need copy instr from 0x2aaf2f87dc  into 0x3fee582048 .

Reply:
Good observation and great detective work. Do all tests pass now? Can we merge this PR?

Reply:
Yes, I think we can merge it. In the CI check, one test failed, but that one has since been fixed, so when this is rebased to the latest, all will pass.

Reply:
Should we instead change the type of reloc? It looks like `INTERNAL_REFERENCE_ENCODED` is intended for an address inside the same function (see reloc-info.h:70) which is not going to always be correct. It seems that perhaps we should use the reloc mode saved in the operand, `j.rmode()`. At first, I tried that and it got a different error because the mode does not get set in the `LoadAddress` case. I added the mode in that case, bit then I am back to the original error. Do we need to go through other places where this `li` is called and set the mode correctly? 

```diff
diff --git a/src/codegen/riscv64/macro-assembler-riscv64.cc b/src/codegen/riscv64/macro-assembler-riscv64.cc
index d7c05c0df9..0e70ee357e 100644
--- a/src/codegen/riscv64/macro-assembler-riscv64.cc
+++ b/src/codegen/riscv64/macro-assembler-riscv64.cc
@@ -1466,7 +1466,7 @@ void TurboAssembler::li(Register rd, Operand j, LiFlags mode) {
   } else if (mode == ADDRESS_LOAD) {
     // We always need the same number of instructions as we may need to patch
     // this code to load another value which may need all 8 instructions.
-    RecordRelocInfo(RelocInfo::INTERNAL_REFERENCE_ENCODED);
+    RecordRelocInfo(j.rmode());
     li_constant(rd, j.immediate());
   } else {  // mode == CONSTANT_SIZE - always emit the same instruction
             // sequence.
@@ -3192,7 +3192,7 @@ void TurboAssembler::Call(Label* target) { BranchAndLink(target); }
 
 void TurboAssembler::LoadAddress(Register dst, Label* target) {
   uint64_t address = jump_address(target);
-  li(dst, address, ADDRESS_LOAD);
+  li(dst, Operand(address, RelocInfo::INTERNAL_REFERENCE_ENCODED), ADDRESS_LOAD);
 }
 
 void TurboAssembler::Push(Smi smi) {
```

Reply:
I have dug a bit deeper and now have a good understanding of the issue here. In *code-generator-riscv64.cc:746*, we try to load an address for the return address from the call. The label it uses is not bound yet, so it starts a chain for the label. A relocation is recorded for this address. Later, that label is bound to an address and this location is updated with the real address. Next, the function gets relocated and the value is updated again, due to the relocation. This all seems correct so far. Then in *wasm-code-manager.cc*, the method `WasmCode::Validate()` goes through and validates all of the relocations. When it sees the relocation I mentioned, it reports an error, because it is labeled as an `INTERNAL_REFERENCE_ENCODED`, but the address is not internal to the WASM function (that's okay, it should not be). I think that it is okay to solve this by changing the relocation type to `EXTERNAL_REFERENCE`. That makes sense in this case (the only place where `LoadAddress` is used), but perhaps not future uses of `LoadAddress`, so I fixed it by passing the mode into `LoadAddress`, and passing `EXTERNAL_REFERENCE` for this use. This fixes the bug and does not cause any new failures. Let me know if you think this makes sense, @luyahan @DrRezaYazdani1.

See my changes here - https://github.com/v8-riscv/v8/commit/c2910e8278969184affd1266438c6b3ac45de30c

Reply:
LGTM